jax/_src/scipy/stats/_core.py

```diff
@@ -0,0 +1,72 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from collections import namedtuple
+from functools import partial
+from typing import Any, List, Optional, Tuple
+
+import scipy
+import numpy as np
+
+from jax import core, jit, lax, vmap
+import jax.numpy as jnp
+from jax._src.numpy.lax_numpy import _check_arraylike
+from jax._src.numpy.reductions import _isscalar
+from jax._src.numpy.util import _wraps
+from jax._src.util import canonicalize_axis
+
+ModeResult = namedtuple("ModeResult", ("mode", "count"))
+
+
+@_wraps(scipy.stats.mode)
+@partial(jit, static_argnames=['axis', 'nan_policy'])
+def mode(x: jnp.ndarray, axis: int = 0, nan_policy: str = "propagate") -> ModeResult:
+
+  def _mode_helper(x: jnp.ndarray) -> Tuple[jnp.ndarray, jnp.ndarray]:
+    "Helper function to return mode and count of a given array."""
+    vals, counts = jnp.unique(x, return_counts=True, size=x.size)
+    return vals[jnp.argmax(counts)], jnp.max(counts)
+
+  _check_arraylike("mode", x)
+  x = jnp.asarray(x)
+  if nan_policy not in {"propagate", "omit", "raise"}:
+    raise ValueError(
+      f"Illegal nan_policy value {nan_policy!r}; expected one of "
+      "{'propoagate', 'omit', 'raise'}"
+    )
+  if nan_policy == "omit":
+    # TODO: return answer without nans included.
+    raise NotImplementedError(
+      f"Logic for `nan_policy` of {nan_policy} is not implemented"
+    )
+  if nan_policy == "raise":
+    # This will always fail with a `nan_policy` of 'raise'. Use a `nan_policy` of
+    # 'propagate' or 'omit' instead. We cannot both JIT compile a function and
+    # have branching logic that depends on its state.
+    _ = core.concrete_or_error(
+      None,
+      x,
+      "In order to best JIT compile `mode`, we cannot know whether `x` contains nans. "
+      "Please check if nans exist in `x` outside of the `mode` function."
+    )
+
+  if x.size == 0:
+    return ModeResult(jnp.array([]), jnp.array([]))
```

Comment:
We should ensure that the dtype of the arrays here match the expected result.

Also, not all empty arrays are the same shape – e.g. the output of mode for an array of shape `[0, 5]` along `axis=1` should probably have shape `[0, 1]`

Reply:
> Also, not all empty arrays are the same shape – e.g. the output of mode for an array of shape [0, 5] along axis=1 should probably have shape [0, 1]

Looking [here](https://github.com/scipy/scipy/blob/4cf21e753cf937d1c6c2d2a0e372fbc1dbbeea81/scipy/stats/_stats_py.py#L411-L412) in `scipy.stats.mode` for loose inspiration into the logic that produces the shape for an empty array.

Check this below, and see if this is what you mean. This is our current behavior - which produces the same result as `scipy.stats.mode`.
```
import numpy as np
from scipy.stats import mode as scipy_mode

from jax._src.scipy.stats._core import mode as jax_mode


cat = np.array([[], []])
cat = cat.reshape((0,2))
cat.shape
#  (0, 2)

scipy_mode(cat, 1)[0].shape
#  (0,)

jax_mode(cat, 1)[0].shape
#  (0,)
```
> We should ensure that the dtype of the arrays here match the expected result.

Did you mean explicitly within the code via an assert statement? If so, this is the only thing I didn't implement in what I'm about to push. I verified here:
```
cat = np.array([1], dtype=np.float16)
scipy_fun(cat, axis=0)[0].dtype
#  dtype('float16')
jax_fun(cat, axis=0)[0].dtype
#  dtype('float16')
```